ckan/lib/uploader.py

```diff
@@ -277,25 +277,25 @@ def upload(self, id, max_size=10):
                 if e.errno != 17:
                     raise
             tmp_filepath = filepath + '~'
-            output_file = open(tmp_filepath, 'wb+')
-            self.upload_file.seek(0)
-            current_size = 0
-            while True:
-                current_size = current_size + 1
-                # MB chunks
-                data = self.upload_file.read(2 ** 20)
-
-                if not data:
-                    break
-                output_file.write(data)
-                if current_size > max_size:
-                    os.remove(tmp_filepath)
-                    raise logic.ValidationError(
-                        {'upload': ['File upload too large']}
-                    )
+            with open(tmp_filepath, 'wb+') as output_file:
+                self.upload_file.seek(0)
+                current_size = 0
+                while True:
+                    current_size = current_size + 1
+                    # MB chunks
+                    data = self.upload_file.read(2 ** 20)
+
+                    if not data:
+                        break
+                    output_file.write(data)
+                    if current_size > max_size:
+                        os.remove(tmp_filepath)
+                        raise logic.ValidationError(
+                            {'upload': ['File upload too large']}
+                        )
+                self.upload_file.close()
```

Comment:
Isn't there a possibility of a file leak here, eg if the 'upload too large' exception is thrown? Shouldn't the code be wrapped in a `try ... finally` block?

Reply:
Sometimes "upload" is type "file" but for small files, ckan uses type StringO (cStringIO.OutputType) which does not implement __enter__/__exit__ so we cannot use "with:". smotornyuk's previous version used "with:" but we had to switch back to ".close()".

Reply:
Yes, I was talking about leaking `self.upload_file`, not the file returned by `open(tmp_filepath)`.

Reply:
i see what you meant, i thought we have `file` object for upload too

Reply:
Yeah, it has sense. @neilmayhew , thanks, i'll check memory leak here(i believe, this piece will cause it, but checks are always valuable:) ) and then i'll update this piece as you suggested